bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/scheduling/Scheduler.java

```diff
@@ -509,13 +516,15 @@ public String toString() {
      * Keep the list of items that should run on same thread.
      * Using a stack to get advantage of the locality.
      */
-    Stack<SchedulerItem> items = new Stack();
+    Stack<SchedulerItem> items = new Stack<>();
 
     /**
      * Indicates this item is already in runnable list/executing or not.
      */
     AtomicBoolean scheduled = new AtomicBoolean(false);
 
+    private final ReentrantLock groupLock = new ReentrantLock();
```

Comment:
Would it be more efficient to keep the a lock in the `ItemGroup`? So there is a minimal chance of locks waiting.

Reply:
+1 for the suggestion

Reply:
Yes it is already inside the ItemGroup. Might confuse since ItemGroup is the inner class of the scheduler.